File: SERVICES/Services1/app/src/main/java/com/javadude/services1/MainActivity5.java
package com.javadude.services1;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.os.RemoteException;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.ProgressBar;

import com.javadude.services2.RemoteService2;
import com.javadude.services2.RemoteService2Reporter;

public class MainActivity5 extends AppCompatActivity {

	private ProgressBar progressBar;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);
		progressBar = (ProgressBar) findViewById(R.id.progressBar);
	}

	private RemoteService2Reporter.Stub reporter = new RemoteService2Reporter.Stub() {
		@Override
		public void report(int i) {
			progressBar.setProgress(i);
		}
	};

	@Override
	protected void onStart() {
		super.onStart();
		Intent intent = new Intent();
		intent.setClassName("com.javadude.services2", "com.javadude.services2.RemoteService2Impl");
		bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE);
	}

	@Override
	protected void onStop() {
		Intent intent = new Intent(this, BoundService2.class);
		stopService(intent);
		super.onStop();
	}

	public void onResetPressed(View view) {
		if (remoteService != null)
			try {
				remoteService.reset();
			} catch (RemoteException e) {
				Log.e(getClass().getSimpleName(), "Cannot reset service count", e);
			}
	}

	private RemoteService2 remoteService;

	private ServiceConnection serviceConnection = new ServiceConnection() {
		@Override
		public void onServiceConnected(ComponentName name, IBinder service) {
			remoteService = RemoteService2.Stub.asInterface(service);
			try {
				remoteService.add(reporter);
			} catch (RemoteException e) {
				Log.e(getClass().getSimpleName(), "Cannot add reporter", e);
			}
		}

		@Override
		public void onServiceDisconnected(ComponentName name) {
			remoteService = null;
		}
	};
}

File: Module - Services/Services1/app/src/main/java/com/javadude/services1/MainActivity5.kt
package com.javadude.services1

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import androidx.appcompat.app.AppCompatActivity
import com.javadude.services2.Person
import com.javadude.services2.RemoteService2
import com.javadude.services2.RemoteService2Reporter
import kotlinx.android.synthetic.main.activity_main.*

// Activity that binds to a remove service (defined in Services2) using AIDL
// NOTE: The AIDL and any parcelables have to be available on both ends!
//       (we could put them in a common module that both depend upon)
// The reporter implements the AIDL reporting interface by extending its generated stub
class MainActivity5 : AppCompatActivity() {

    private val reporter = object : RemoteService2Reporter.Stub() {
        override fun report(people: List<Person>, i: Int) {
            runOnUiThread {
                progressBar.progress = i
                textView.text = people.joinToString("\n") {
                    "${it.name}: ${it.age}"
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        reset_button.setOnClickListener {
            binder?.reset()
        }

        other_activity_button.setOnClickListener {
            startActivity(Intent(this, MainActivity3a::class.java))
        }
    }

    override fun onStart() {
        super.onStart()
        val intent = Intent().apply {
            // Need to specify a different application's ID here
            setClassName("com.javadude.services2", "com.javadude.services2.RemoteService2Impl")
        }
        bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE)
    }

    override fun onStop() {
        binder?.remove(reporter)
        unbindService(serviceConnection)
        super.onStop()
    }


    private var binder : RemoteService2? = null

    private val serviceConnection = object : ServiceConnection {
        override fun onServiceDisconnected(p0: ComponentName?) {
            this@MainActivity5.binder = null
        }

        override fun onServiceConnected(p0: ComponentName?, binder: IBinder?) {
            this@MainActivity5.binder = RemoteService2.Stub.asInterface(binder)
            this@MainActivity5.binder!!.add(reporter)
        }
    }
}

